src/mscorlib/src/System/Diagnostics/Stacktrace.cs

```diff
@@ -75,7 +85,56 @@ public StackFrameHelper(bool fNeedFileLineColInfo, Thread target)
             // override it.
             iFrameCount = 0;
 
-            fNeedFileInfo = fNeedFileLineColInfo;
+            fNeedFileInfo = fNeedFileLine;
+        }
+
+        //
+        // Initializes the rgFilename, rgiLineNumber and rgiColumnNumber fields using the portable PDB reader.
+        //
+        internal void InitializeSourceInfo()
+        {
+            if (!fNeedFileInfo) {
+                return;
+            }
+
+            if (s_symbolsType == null) {
+                try {
+                    Assembly metadataAssembly = Assembly.Load("System.Diagnostics.StackTrace.Symbols, Version=1.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a");
+                    if (metadataAssembly == null) {
+                        return;
+                    }
+                    s_symbolsType = metadataAssembly.GetType("System.Diagnostics.StackTrace.Symbols");
+                    if (s_symbolsType == null) {
+                        return;
+                    }
+                }
+                catch {
+                    return;
+                }
+            }
+
+            for (int index = 0; index < iFrameCount; index++) {
+                // If there was some reason not to try get get the symbols from the portable PDB reader like the module was
+                // ENC or the source/line info was already retrieved, the assembly path is null.
+                if (rgAssemblyPath[index] != null || rgInMemorySymbols[index] != IntPtr.Zero) {
+                    try {
+                        object[] parameters = new object[8];
+                        parameters[0] = rgAssemblyPath[index];
+                        parameters[1] = rgInMemorySymbols[index];
+                        parameters[2] = rgiInMemorySymbolsSize[index];
+                        parameters[3] = rgiMethodToken[index];
+                        parameters[4] = rgiILOffset[index];
+
+                        s_symbolsType.InvokeMember("GetSourceLineInfo", BindingFlags.InvokeMethod, null, null, parameters);
```

Comment:
It is common to create to the method you need to callback to, to avoid paying for the reflection overhead each time - it makes the code look better as well.